### ISSUE TITLE
vkCmdExecuteCommands complains about host sync for secondary cmdbuffer pools

### ISSUE THREAD

**Issue by pixeljetstream** [KhronosGroup/Vulkan-ValidationLayers#376]
Following error is thrown for vkCmdExecuteCommands:
[ UNASSIGNED-Threading-MultipleThreads ] Object: 0xad (Type = 25) | THREADING ERROR : object of type VkCommandPool is simultaneously used in thread 0x4de0 and thread 0x481c

the scenario is as is
primary thread -> vkCmdExecuteCommands ( primary_cmd,  1, &secondary_cmd )
secondary thread -> generates new secondary_cmds in parallel using secondary_pool
primary_cmd has its own pool.

As the secondary commandbuffers are not being modified during vkCmdExecuteCommands, their pools should not be locked during vkCmdExecuteCommands.

**Comment by jeffbolznv** [comment]
> patching, building shadow command buffers when you can't patch inplace but need to do something, etc

Shouldn't these happen in the primary's pool?

> it's perfectly reasonable for its /reads/ to need to be synchronized wrt writes made against the same pool.

Reading the memory already allocated to the secondary command buffer shouldn't require a read lock on the pool.

**Comment by chrisforbes** [comment]
The concrete case I'm thinking of is CmdExecuteCommands racing with
ResetCommandPool.

**Comment by jzulauf-lunarg** [comment]
From my reading of the error message it can't be that both sets of command involved have separate pools.  The error is noting that commands that impact the *same* command pool is being used by more than one thread.

**Comment by jzulauf-lunarg** [comment]
@pixeljetstream  -- did my comments make sense here, or am I mis-understanding the issue?

**Comment by pixeljetstream** [comment]
To me unfortunately they made no sense. The secondary cmd buffers are not modified, they are not the problem.
Their pool is used. But I still think what I do should be legal as cmdexecutecommands is not an operation that modifies. Yes a reset would nuke everything, but wouldn't that also nuke the cmdbuffers first...

**Comment by jzulauf-lunarg** [comment]
@pixeljetstream  -- my misunderstanding of the secondary command buffer updates. 

Looking at the validation message, the threading layer is seeing the *same* pool object in use on two threads -- so unless there is a layer *above* validation that is injecting additional traffic (and barring a bug in thread validation or unique objects), somehow two application threads are making Vulkan API calls that use same command pool handle (or modify command buffers allocated from the same pool) . 

c.f. Implicit Externally Synchronized Parameters in section 2.6

> all commandBuffer parameters that need to be externally synchronized imply that the commandPool that was passed in when creating that command buffer also needs to be externally synchronized. 

Have you dumped the pool handles so we can know which pool is getting double used?

**Comment by jzulauf-lunarg** [comment]
@pixeljetstream -- note the rewrite above...

**Comment by pixeljetstream** [comment]
Thanks for the clarification, was unaware of this implicit rule. That means validation is correct.

**Comment by jzulauf-lunarg** [comment]
Had to dig into the spec to find it :) -- closing.

**Comment by jeffbolznv** [comment]
> However, I think @jeffbolznv and @chrisforbes are right. You must externally synchronize CmdExecuteCommands against both the primary and secondary pools.

This was not my position. I don't think it is, or should be, necessary for the secondary command buffers to be externally synchronized.

> In-place patching of a secondary only makes sense in the secondary's pool.

In-place patching shouldn't require touching the pool, it only requires writes to the memory already allocated to the secondary command buffers.

> c.f. "Implicit Externally Synchronized Parameters" in section 2.6
>>     all commandBuffer parameters that need to be externally synchronized imply that the commandPool that was passed in when creating that command buffer also needs to be externally synchronized.

This is irrelevant because the secondary command buffers are not externally synchronized.

> Each element of the pCommandBuffers parameter in vkFreeCommandBuffers

That is vkFreeCommandBuffers, not vkCmdExecuteCommands.

Reopening. IMO @pixeljetstream's original issue is valid.

**Comment by jzulauf-lunarg** [comment]
@jeffbolznv  D'oh... correct re: Free vs. Exec

**Comment by chrisforbes** [comment]
I agree with @jeffbolznv that validation is operating outside the spec here. I'm still not convinced secondary command buffers _shouldn't be_ externally synchronized, but that's a spec question :)

**Comment by jzulauf-lunarg** [comment]
The section of the spec we are discussing is autogenerated from the vk.xml, the difference being that for vkFreeCommandBuffers pCommandBufers is external_sync=true whereas for vkCmdExecuteCommands it is not.  So it appears to be intentional.  The validation layers will honor the below clause in the *negative*.  If a commandBuffer does not require external synchronization, the command Pool will *not* be treated as implicitly requiring external synchronization.  Practically this means that those command *buffers* will be marked as read accessed (and thus report conflicts with any call with *write* access to that same buffer, but the associated command pool will be ignored -- treated as not accessed.

> all commandBuffer parameters that need to be externally synchronized imply that the commandPool that was passed in when creating that command buffer also needs to be externally synchronized.


Certainly for "Reset" and "Destroy" operations on command pools the race needs to be caught, and the way to look at it is  that effectively these have implicit external sync requirements on the *contents* of the pool, i.e. every command buffer within it, s.t. conflicts can be caught even when the command pool itself  is not considered accessed.

**Comment by jzulauf-lunarg** [comment]
@pixeljetstream  -- branch at PR 424 designed to address this, but don't have a test case.

@jeffbolznv @chrisforbes  -- FYI proposed change should still catch races vs. Reset and Destroy CommandPool using a new mechanism -- the FreeCommandBuffers case was already covered and remains so.
